hub/hub.go

```diff
@@ -4,80 +4,186 @@ import (
+// WithPublisherJWTConfig sets the JWT key and algorithm to use.
+func WithPublisherJWTConfig(key []byte, signingMethod jwt.SigningMethod) Option {
+	return func(h *Hub) {
+		h.publisherJWTConfig = &jwtConfig{key, signingMethod}
+	}
+}
+
+// WithSubscriberJWTConfig sets the JWT key and algorithm to use.
+func WithSubscriberJWTConfig(key []byte, signingMethod jwt.SigningMethod) Option {
+	return func(h *Hub) {
+		h.subscriberJWTConfig = &jwtConfig{key, signingMethod}
+	}
 }
 
-// NewHubWithTransport creates a hub.
-func NewHubWithTransport(v *viper.Viper, t Transport, logger Logger, tss *TopicSelectorStore) *Hub {
-	return &Hub{
-		v,
-		logger,
-		t,
-		nil,
-		nil,
-		tss,
-		NewMetrics(),
+// WithPublishOrigins sets the origins allowed to publish updates.
+func WithPublishOrigins(origins []string) Option {
+	return func(h *Hub) {
+		h.publishOrigins = origins
 	}
 }
 
-// Start is an helper method to start the Mercure Hub.
-func Start() {
-	h, err := NewHub(viper.GetViper())
+// WithTransportURL sets the transport to use by parsing the provided URL.
+func WithTransportURL(tu string) Option {
+	u, err := url.Parse(tu)
 	if err != nil {
-		log.Fatalln(err)
+		log.Panic(fmt.Errorf("transport_url: %w", err))
 	}
 
-	defer func() {
-		if err = h.Stop(); err != nil {
-			log.Fatalln(err)
+	return func(h *Hub) {
+		h.transportURL = u
+	}
+}
+
+type jwtConfig struct {
+	key           []byte
+	signingMethod jwt.SigningMethod
+}
+
+// Hub stores channels with clients currently subscribed and allows to dispatch updates.
+type Hub struct {
+	anonymous           bool
+	debug               bool
+	demo                bool
+	subscriptions       bool
+	logger              Logger
+	writeTimeout        time.Duration
+	dispatchTimeout     time.Duration
+	heartbeat           time.Duration
+	publisherJWTConfig  *jwtConfig
+	subscriberJWTConfig *jwtConfig
+	publishOrigins      []string
+	transportURL        *url.URL
+	transport           Transport
+	metrics             *Metrics
+	topicSelectorStore  *TopicSelectorStore
+
+	// Deprecated: use the Caddy server module or the standalone library instead.
+	config        *viper.Viper
+	server        *http.Server
+	metricsServer *http.Server
+}
+
+func New(options ...Option) *Hub {
+	h := &Hub{
+		topicSelectorStore: NewTopicSelectorStore(),
+		writeTimeout:       600 * time.Second,
+	}
+
+	for _, o := range options {
+		o(h)
+	}
+
+	if h.logger == nil {
+		var (
+			l   Logger
+			err error
+		)
+		if h.debug {
+			l, err = zap.NewDevelopment()
+		} else {
+			l, err = zap.NewProduction()
 		}
-	}()
 
-	h.Serve()
+		if err != nil {
+			log.Panic(err)
+		}
+
+		h.logger = l
+	}
+
+	if h.transportURL == nil {
+		h.transportURL = &url.URL{Scheme: "local"}
+	}
+
+	t, err := newTransport(h.transportURL, h.logger)
+	if err != nil {
+		log.Panic(err)
+	}
+	h.transport = t
+
+	return h
+}
+
+// Stop stops disconnect all connected clients.
+func (h *Hub) Stop() error {
+	return h.transport.Close()
```

Comment:
If we consider the transport an external dependency, the `Close` method should be called by the `main` function.

I would actually remove the `Close` method from the interface entirely, because it shouldn't be the Hub's concern.

Then in the main function:

```
transport, err := createTransportFromURL(u)
if err != nil {
    panic(err)
}

// This is only required so that createTransportFromURL can return an interface
// if in the main someone constructs a concrete transport type
// then the type assertion won't be necessary
if closer, ok := transport.(interface{ Close() error}); ok {
    defer closer.Close()
}

hub := hub.New(WithTransport(transport))
```